GDJS/Runtime/AsyncTasksManager.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * GDevelop JS Platform
+ * Copyright 2013-2022 Florian Rival (Florian.Rival@gmail.com). All rights reserved.
+ * This project is released under the MIT License.
+ */
+namespace gdjs {
+  /**
+   * A manager for all asynchronous tasks.
+   * @see {@link RuntimeScene.getAsyncTasksManager}.
+   */
+  export class AsyncTasksManager {
+    /**
+     * Maps a task to the callback to be executed once it is finished.
+     */
+    private tasks = new Map<AsyncTask, (runtimeScene: RuntimeScene) => void>();
+
+    /**
+     * Run all pending asynchronous tasks.
+     */
+    processTasks(runtimeScene: RuntimeScene): void {
+      for (const task of this.tasks.keys()) {
+        if (task.update(runtimeScene)) {
+          // The task has finished, run the callback and remove it.
+          this.tasks.get(task)!(runtimeScene);
+          this.tasks.delete(task);
+        }
+      }
+    }
+
+    /**
+     * Adds a task to be processed between frames and a callback for when it is done to the manager.
+     * @param task The {@link AsyncTask} to run.
+     * @param then The callback to execute once the task is finished.
+     */
+    addTask(task: AsyncTask, then: (runtimeScene: RuntimeScene) => void): void {
+      this.tasks.set(task, then);
+    }
+  }
+
+  /**
+   * An asynchronous task to be run between frames.
+   */
+  export abstract class AsyncTask {
+    /**
+     * Called every frame where the scene is active.
+     * @param runtimeScene - The scene the task runs on.
+     * @return True if the task is finished, false if it needs to continue running.
+     */
+    abstract update(runtimeScene: RuntimeScene): boolean;
+  }
+
+  /**
+   * A task that resolves with a promise.
+   */
+  export class PromiseTask extends AsyncTask {
+    private isResolved: boolean = false;
+
+    constructor(promise: Promise<any>) {
+      super();
+      promise
+        .then(() => {
+          this.isResolved = true;
+        })
+        .catch(() => {});
```

Comment:
This means a failing promise, like a failed network request, will never execute the following actions. This might be fine, but we need to make it explicit that in this case, promise must NEVER fail (even if there is an error, they should catch it and expose it in some way to the events, probably using a variable). And if a promise fail, we should probably log it or crash the game.

Reply:
Maybe we can ensure that the promise error is handled by requiring it as part of the `PromiseTask` like that:
![image](https://user-images.githubusercontent.com/19349038/153729133-b9717125-e4c5-4e62-8e67-6804387f2bd8.png)

Also, I'm unsure whether a promise failure should cancel the execution of a callback, since that would prevent users from doing error handling, but still executing it would result in either unexpected scenarios for users where the data they waited for does not exist even after the wait in the case of for example a network request

Reply:
That's a design decision basically to do:
- Promise (and async/await) in JS can either be resolved or throw (be "rejected").
- But in our abstraction of asynchronous actions, there are no differences between the success or the failure. There is no "exception" or "sad path", an asynchronous action is supposed to always resolve after a bit of time.

So we must either:
- Consider that promise will never fail, and crash if it's the case.
- Or consider that if they fail, we automatically report this somehow... but how? 

In my view, it's the responsibility of the action implementation to do that, no? If a network request is failing, it's your job to catch it and report it (either in an error or result variable, with the error code - or whatever as long as you give a way to check the error, maybe another condition specialised for this?).
So a AsyncTask should just resolve and the error handling should be done in the implementation of the action itself (a wait can not fail, a request can fail but that's just another error like you could get a 4xx or 5xx, etc...)?

Reply:
> I'm unsure whether a promise failure should cancel the execution of a callback

I think it should not. Something launched should always finish at some point (and you might even want to set up a timeout or have a default timeout to avoid waiting for 10 minutes for a request that will never resolve).
But if you have an async action, the rest should always be executed at some point.

Only exception: you stopped the scene in the meantime.

Reply:
I opted for expecting all promises to never fail. If they still do and did not add their own `.catch`, I do not crash the game, but log an error message in the console. Either way, the callback gets called.